src/devices/Mcp3428/Mcp342x.cs

```diff
@@ -0,0 +1,387 @@
+            {
+                errorList?.Add($"Channel update failed from {_lastChannel} to {channel}");
+                ok = false;
+            }
+            if (Resolution != resolution)
+            {
+                errorList?.Add($"Resolution update failed from {Resolution} to {resolution}");
+                ok = false;
+            }
+            if (mode != Mode)
+            {
+                errorList?.Add($"Mode update failed from {Mode} to {mode}");
+                ok = false;
+            }
+            if (PGAGain != pgaGain)
+            {
+                errorList?.Add($"PGAGain update failed from {PGAGain} to {pgaGain}");
+                ok = false;
+            }
+
+            if (!ok)
+            {   // Only use console on error
+                errorList?.Add($"Sent config byte {conf:X}, received {LastConfigByte:X}");
+            }
+            return ok;
+        }
+
+        protected int WaitTime => (int)(1000.0 / Helpers.UpdateFrequency(Resolution));
+
+        public int ChannelCount { get; }
+
+        protected ConversionResult LastConversion
+        {
+            get
+            {
+                return _lastConversion;
+            }
+            set
+            {
+                _lastConversion = value;
+                OnConversion?.Invoke(this, _lastConversion);
+            }
+        }
+
+        public event EventHandler<ConversionResult> OnConversion;
+
+        protected void WriteConfig(byte configByte)
+        {
+            _i2cDevice.WriteByte(configByte);
+        }
+
+        private static int _asyncThreshold = 20;
+
+        /// <summary>
+        /// Sets the lower time limit in ms. If the current configuration reads data faster the synchronous API is used.
+        /// This can save some overhead.
+        /// </summary>
+        /// <remarks>
+        /// The default configuration is 20ms. This means that only in 16 bit resolution read waits asynchronously.
+        /// Setting it to 0 or lower disables the function.
+        /// </remarks>
+        /// <param name="thresh">Time limit in ms. Default: 20ms</param>
+        /// <autogeneratedoc />
+        public static void SetAsyncThreshold(int thresh) { _asyncThreshold = thresh; }
+
+        /// <summary>
+        /// One-shot read as an asynchronous operation. Initiates read and waits for it to finish.
+        /// Async API was required as reading in 16bit resolution can take more than 60ms.
+        /// </summary>
+        /// <param name="channel">The channel.</param>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns>ValueTask.</returns>
+        /// <exception cref="System.IO.IOException">
+        /// Device is not in One-Shot mode
+        /// or
+        /// ADC Conversion was not ready after {tries}
+        /// </exception>
+        /// <autogeneratedoc />
+        protected ValueTask OneShotReadAsync(int channel = -1, CancellationToken cancellationToken = default)
```

Comment:
should methods taking `channel` throw when invalid value is used?

Reply:
I think it does in the bit setter function.